cmd/pki.go

```diff
@@ -0,0 +1,354 @@
+	
+	The parameters are to be specified in the configuration file or inline.
+	
+	The master node has probably an IP address as hostname, if you run your instance from a box without DNS indexing.
+	
+	The client nodes don't need a publicly known hostname, but you may assign them a name that suits your use (e.g. your logs may see "client_x successfully connected to API"")`,
+
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		config.LoadConfig(configPath)
+
+		if serverID == "" {
+			serverID = config.Config.Server.IP
+		}
+
+	},
+
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("Setting up PKI at " + certificateDirectory)
+		pkiSetup(organisationSubject, nNodes, certDuration, ecdsaCurve, certificateDirectory, serverID)
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(pkiCmd)
+
+	pkiCmd.Flags().UintVarP(&nNodes, "nodes", "n", 1, "Number of nodes to create certificates for")
+	pkiCmd.Flags().StringVarP(&serverID, "server", "@", "", "Server address, with IP or DNS name")
+	pkiCmd.Flags().StringVarP(&ecdsaCurve, "ec", "e", "P256", "Elliptic curve to use for certificates. Accepted values : P224, P256 (recommended), P384 and P521")
+	pkiCmd.Flags().BoolVarP(&keepPrivCA, "keepca", "k", false, "Indicate whether to keep CA private key (i.e. written to disk)")
+	pkiCmd.Flags().DurationVarP(&certDuration, "duration", "d", 365*24*time.Hour, "Duration for certificate. Default is a year")
+	pkiCmd.Flags().StringVarP(&organisationSubject, "org", "o", "Netm4ul", "Organisation Subject to use in server certificate")
+	pkiCmd.Flags().StringVarP(&certificateDirectory, "dir", "t", "./certificates", "Local directory to store PKI certs and keys")
+
+	viper.SetDefault("ec", "P256")
+	viper.SetDefault("keepca", false)
+	viper.SetDefault("duration", 365*24*time.Hour)
+	viper.SetDefault("subject", "Netm4ul")
+	viper.SetDefault("dir", "./certificates")
+
+	pkiCmd.MarkFlagRequired("nodes")
+}
+
+// Extract public key from private
+func publicKey(priv interface{}) interface{} {
+	switch k := priv.(type) {
+	/*case *rsa.PrivateKey:
+	return &k.PublicKey*/
+	case *ecdsa.PrivateKey:
+		return &k.PublicKey
+	default:
+		return nil
+	}
+}
+
+func pemBlockForKey(priv interface{}) *pem.Block {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}
+	case *ecdsa.PrivateKey:
+		b, err := x509.MarshalECPrivateKey(k)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Unable to marshal ECDSA private key: %v", err)
+			os.Exit(2)
+		}
+		return &pem.Block{Type: "EC PRIVATE KEY", Bytes: b}
+	default:
+		return nil
+	}
+}
+
+/*
+func get_formatted_time() string {
+	now := time.Now()
+
+	nowH, nowM, nowS := now.Clock()
+
+	return fmt.Sprintf("%s %d %d:%d:%d %d", now.Month(), now.Day(), nowH, nowM, nowS, now.Year())
+}
+*/
+
+// Generates a ecdsa private key from a cryptographically secure prng
+func generatePrivateKey(ecdsaCurve string) interface{} {
+	var priv interface{}
```

Comment:
Is the `interface{}` type required here ?
Maybe [ecdsa.PrivateKey](https://golang.org/pkg/crypto/ecdsa/#PrivateKey) could be used?

Reply:
Like discussed about possible but undesired RSA support, I will have a look. Let's toss the support and go full ec.